rapids-cmake/cpm/detail/pinning_write_file.cmake

```
@@ -0,0 +1,257 @@
+                    WORKING_DIRECTORY ${source_dir}
+                    ERROR_QUIET
+                    OUTPUT_VARIABLE _rapids_commit_stack
+                    OUTPUT_STRIP_TRAILING_WHITESPACE)
+    # The last entry in the output that has "* [" is our commit
+    #
+    # Find that line and convert the `* [short-sha1] Commit Message` to a list that is ` *
+    # ;short-sha1;Commit Message` and extract the short sha1
+    string(FIND "${_rapids_commit_stack}" "* [" position REVERSE)
+    if(position LESS 0)
+      # No changes to the repo so use the `HEAD` keyword
+      set(short_sha HEAD)
+    else()
+      string(SUBSTRING "${_rapids_commit_stack}" ${position} -1 _rapids_commit_stack)
+      string(REGEX REPLACE "(\\[|\\])" ";" _rapids_commit_stack "${_rapids_commit_stack}")
+      list(GET _rapids_commit_stack 1 short_sha)
+    endif()
+
+    # Convert from the short sha1 ( could be keyword `HEAD` ) to a full SHA1
+    execute_process(COMMAND ${GIT_EXECUTABLE} rev-parse ${short_sha}
+                    WORKING_DIRECTORY ${source_dir}
+                    ERROR_QUIET
+                    OUTPUT_VARIABLE _RAPIDS_SHA
+                    OUTPUT_STRIP_TRAILING_WHITESPACE)
+  endif()
+  # Only set the provided variables if we extracted the information
+  if(_RAPIDS_URL)
+    set(${git_url_var} "${_RAPIDS_URL}" PARENT_SCOPE)
+  endif()
+  if(_RAPIDS_SHA)
+    set(${git_sha_var} "${_RAPIDS_SHA}" PARENT_SCOPE)
+  endif()
+
+endfunction()
+
+#[=======================================================================[.rst:
+rapids_cpm_pinning_add_json_entry
+---------------------------------
+
+.. versionadded:: v24.04.00
+
+#]=======================================================================]
+function(rapids_cpm_pinning_add_json_entry json_var package_name url_var sha_var)
+  include("${rapids-cmake-dir}/cpm/detail/get_default_json.cmake")
+  include("${rapids-cmake-dir}/cpm/detail/get_override_json.cmake")
+  get_default_json(${package_name} json_data)
+  get_override_json(${package_name} override_json_data)
+
+  set(url_string)
+  set(sha_string)
+  if(${url_var})
+    string(CONFIGURE [=["git_url": "${${url_var}}",]=] url_string)
+  endif()
+  if(${sha_var})
+    string(CONFIGURE [=["git_tag": "${${sha_var}}",]=] sha_string)
+  endif()
+  # We start with a default template, and only add members that don't exist
+  string(CONFIGURE [=[{
+  "version": "${CPM_PACKAGE_${package_name}_VERSION}",
+  ${url_string}
+  ${sha_string}
+  "git_shallow": "false",
+  "always_download": "true"
+  }]=]
+                   pinned_json_entry)
+
+  # Insert a json key and value
+  #
+  function(rapids_cpm_pinning_create_and_set_member json_blob_var key value)
+
+    # if the first char in value is { or [ we don't need quotes
+    if(NOT value MATCHES "^(\\{|\\[)")
+      set(value "\"${value}\"")
+    endif()
+    # We need to quote 'value' so that it is a valid json element. Only ones not allowed to be
+    # unquted are numbers by the json spec
```

Strings must be quoted. Numbers like float `1.23`, or integer `456`, booleans like `true`, null values `null`, or arrays `[1, 2, 3]` aren't quoted. I assume arrays aren't relevant here, but booleans in particular might be important?
```suggestion
    # unquoted are numbers, booleans, or null by the JSON spec
```